components/tls/tls-client-async/src/lib.rs

```
@@ -74,165 +72,163 @@ impl Future for ConnectionFuture {
+
+        // Runs both the tx and rx halves of the connection to completion.
+        // This loop does not terminate until the *SERVER* closes the connection and
+        // we've processed all received data. If an error occurs, the `TlsConnection`
+        // channels will be closed and the error will be returned from this future.
+        'conn: loop {
+            // Write all pending TLS data to the server.
+            if client.wants_write() && !client_closed {
+                #[cfg(feature = "tracing")]
+                trace!("client wants to write");
+                while client.wants_write() {
+                    let _sent = client.write_tls_async(&mut server_tx).await?;
+                    #[cfg(feature = "tracing")]
+                    trace!("sent {} tls bytes to server", _sent);
+                }
+                server_tx.flush().await?;
+            }
 
-        'outer: loop {
-            select_biased! {
-                read_res = &mut rx_tls_fut => {
-                    let received = read_res?;
+            // Forward received plaintext to `TlsConnection`.
+            while !client.plaintext_is_empty() {
+                let read = client.read_plaintext(&mut rx_buf)?;
+                recv.extend(&rx_buf[..read]);
+                // Ignore if the receiver has hung up.
+                _ = rx_sender
+                    .send(Ok(Bytes::copy_from_slice(&rx_buf[..read])))
+                    .await;
+                #[cfg(feature = "tracing")]
+                trace!("forwarded {} plaintext bytes to conn", read);
+            }
+
+            if !client.is_handshaking() && !handshake_done {
+                #[cfg(feature = "tracing")]
+                debug!("handshake complete");
+                handshake_done = true;
+                // Start reading application data that needs to be transmitted from the `TlsConnection`.
+                tx_recv_fut = tx_receiver.next().fuse();
```

Seems like with the current logic it is possible to silently lose some tx `data`.

1. Consider if some tx `data` is available at this point, then `tx_recv_fut` should resolve to that `data`.

2. Consider if in the `select_biased` loop below only the `rx_tls_fut` branch will be executed because there is some incoming data from the server (`select_biased!` executes only one branch iiuc). This means that the `tx_recv_fut` branch will **not** be executed and the `data` will **not** be sent.

3. In the next loop iteration `tx_recv_fut` on this line will be set to a future which resolves to something else, e.g. `None`. This will result in `data` not getting sent.